test/membrane/error.spec.js

```diff
@@ -0,0 +1,212 @@
+                
+                class Foo {
+                    constructor() {
+                        throw errorObj;
+                    }
+                }
+
+                try {
+                    new Foo()
+                } catch(e) {
+                    expect(e.foo).toBe('bar')
+                    expect(e).toBe(errorObj)
+                }
+            `)
+})
+
+it('non-error objects thrown in Promise', (done) => {
+    const evalScript = createSecureEnvironment(undefined, { done, expect })
+    evalScript(`
+                const error = { foo: 'bar' }
+                const p = new Promise(() => {
+                    throw error
+                })
+                p.catch((e) => {
+                    expect(e.foo).toBe('bar')
+                    expect(e).toBe(error)
+                    done()
+                })
+            `)
+})
+
+it('unhandled promise rejections with non-error objects', (done) => {
+    const evalScript = createSecureEnvironment(undefined, { done, expect })
+    evalScript(`
+        const errorObj = { foo: 'bar' }
+        window.addEventListener("unhandledrejection", event => {
+            expect(event.reason).toBe(errorObj)
+            expect(event.reason.foo).toBe('bar')
+            done()
+        });
+
+        new Promise((resolve, reject) => {
+            throw errorObj
+        });
+    `)
+})
+
+it('non-error objects thrown in blue functions do not leak', () => {
+    function foo() {
+        throw { foo: 'bar' }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { foo, expect })
+    evalScript(`
+                try {
+                    foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in blue constructors do not leak', () => {
+    class Foo {
+        constructor() {
+            throw { foo: 'bar' }
+        }
+    }
+    
+    const evalScript = createSecureEnvironment(undefined, { Foo, expect })
+    evalScript(`
+                try {
+                    new Foo()
+                } catch (e) {
+                    // blue error own properties do not leak
+                    expect(e.foo).toBe(undefined)
+                }
+            `)
+})
+
+it('non-error objects thrown in red functions do not leak', () => {
+    let fn
+    function save(arg) {
+        fn = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
+    evalScript(`
+        function foo() {
+            throw { foo: 'bar' }
+        }
+    `)
+
+    try {
+        fn()
+    } catch(e) {
+        expect(e.foo).toBe(undefined);
+    }
+})
+
+it('non-error objects thrown in red consturctors do not leak', () => {
+    let ctor
+    function save(arg) {
+        ctor = arg
+    }
+
+    const evalScript = createSecureEnvironment(undefined, { save })
```

Comment:
why pass save?

Reply:
same as above... it needs to be passed, and it needs to be invoked with `Foo` ref as the first argument... but it happens to pass... @dejang was lucky! jajaja

Reply:
missed it :) added now